core/src/main/scala/akka/kafka/internal/BaseSingleSourceLogic.scala

```diff
@@ -108,17 +108,22 @@ import scala.concurrent.{ExecutionContext, Future}
     consumerActor.tell(KafkaConsumerActor.Internal.RequestMessages(requestId, tps), sourceActor.ref)
   }
 
-  setHandler(shape.out, new OutHandler {
-    override def onPull(): Unit = pump()
-    override def onDownstreamFinish(cause: Throwable): Unit =
-      performShutdown()
-  })
+  setHandler(
+    shape.out,
+    new OutHandler {
+      override def onPull(): Unit = pump()
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        super.onDownstreamFinish(cause)
```

Comment:
This will by default `thisStage.cancelStage`. Will that work well together with subclass implementations such as in `SingleSourceLogic`, which is using `setKeepGoing(true)` and `complete(shape.out)`?

Reply:
I'm not sure. Maybe nothing we want to touch without trying it harder.